tests/unit/server/bundle_manager/__init__.py

```diff
@@ -0,0 +1,216 @@
+}
+
+BASE_METADATA_DATASET_BUNDLE = {
+    "description": "",
+    "name": "dataset-1",
+    "created": 1495784349,
+    "failure_message": "",
+    "tags": [],
+    "license": "",
+    "source_url": "",
+}
+
+
+class BaseBundleManagerTest(unittest.TestCase):
+    """
+    Base class for BundleManager tests with a CodaLab Manager hitting a real, in-memory database.
+    """
+
+    def setUp(self):
+        self.codalab_manager = CodaLabManager()
+        self.codalab_manager.config['server']['class'] = 'SQLiteModel'
+        self.bundle_manager = BundleManager(self.codalab_manager)
+        self.user_id = generate_uuid()
+        self.bundle_manager._model.add_user(
+            "codalab",
+            "noreply@codalab.org",
+            "Test",
+            "User",
+            "password",
+            "Stanford",
+            user_id=self.user_id,
+        )
+
+    def create_make_bundle(self, state=State.MAKING):
+        bundle = MakeBundle.construct(
+            targets=[],
+            command='',
+            metadata=BASE_METADATA_MAKE_BUNDLE,
+            owner_id=self.user_id,
+            uuid=generate_uuid(),
+            state=state,
+        )
+        return bundle
+
+    def save_bundle(self, bundle):
+        self.bundle_manager._model.save_bundle(bundle)
+
+    def read_bundle(self, bundle, extra_path=""):
+        return open(
+            os.path.join(
+                self.codalab_manager.bundle_store().get_bundle_location(bundle.uuid), extra_path
+            ),
+            "r",
+        )
+
+    def create_run_bundle(self, state=State.CREATED, metadata=None):
+        bundle = RunBundle.construct(
+            targets=[],
+            command='',
+            metadata=dict(BASE_METADATA, **(metadata or {})),
+            owner_id=self.user_id,
+            uuid=generate_uuid(),
+            state=state,
+        )
+        return bundle
+
+    def create_bundle_single_dep(
+        self, parent_state=State.READY, bundle_state=State.CREATED, bundle_type=RunBundle
+    ):
+        parent = self.create_run_bundle(parent_state)
+        with open(self.codalab_manager.bundle_store().get_bundle_location(parent.uuid), "w+") as f:
+            f.write("hello world")
+        bundle = (
+            self.create_run_bundle(bundle_state)
+            if bundle_type == RunBundle
+            else self.create_make_bundle(bundle_state)
+        )
+        bundle.dependencies = [
+            Dependency(
+                {
+                    "parent_uuid": parent.uuid,
+                    "parent_path": "",
+                    "child_uuid": bundle.uuid,
+                    "child_path": "src",
+                }
+            )
+        ]
+        return bundle, parent
+
+    def create_bundle_two_deps(self):
+        parent1 = self.create_run_bundle(state=State.READY)
+        with open(self.codalab_manager.bundle_store().get_bundle_location(parent1.uuid), "w+") as f:
+            f.write("hello world 1")
```

Comment:
Can we refactor and push adding contents into `create_run_bundle` to eliminate duplicate code?

Reply:
I'd rather not -- this `open(...)` code is only called within the reused / helper functions three times, not in actual test cases. Additionally, I can't think of a good, general way to add contents into create_run_bundle that would also take into account, say, creating a bundle directory. Finally, `create_run_bundle` doesn't itself save bundles to the database, so it might be a bit confusing if it constructs a bundle but then also persists data to the filesystem.

Reply:
What about creating another function `create_run_bundle_with_contents` or something?  It's three times now, but it seems likely that the number will grow as we add more tests.

Reply:
Done